product_merge/wizards/product_merge_wizard.py

```diff
@@ -0,0 +1,207 @@
+                            "Product '%(product)s' has multiple variants.",
+                            product=product.name,
+                        )
+                    )
+
+    @api.constrains("product_ids")
+    def _check_product_types(self):
+        for wizard in self:
+            if len(wizard.product_ids) > 1:
+                types = wizard.product_ids.mapped("type")
+                if len(set(types)) > 1:
+                    raise ValidationError(
+                        _(
+                            "All products to merge must be of the same type "
+                            "(e.g., consumable, service, or storable)."
+                        )
+                    )
+
+    @api.model
+    def default_get(self, fields_list):
+        res = super().default_get(fields_list)
+        if self.env.context.get(
+            "active_model"
+        ) != "product.template" or not self.env.context.get("active_ids"):
+            return res
+        res["product_ids"] = [Command.set(self.env.context["active_ids"])]
+        return res
+
+    @api.model
+    def _get_merge_field_spec(self):
+        """This method defines the merge strategy for each field. By default, we choose
+        to force the target value, but this method can be inherited to change the
+        behavior of the merge.
+        See the _adjust_merged_values_orm method documentation in OpenUpgradeLib for
+        more details."""
+        return {
+            field_name: "target" for field_name in self.product_tmpl_id._fields.keys()
+        }
+
+    def action_merge_products(self):
+        self.ensure_one()
+        self._check_minimum_products()
+        self._check_products_max_one_variant()
+        self.product_tmpl_id.with_context(product_merge=True).write(
+            {
+                "attribute_line_ids": [
+                    Command.create(
+                        {
+                            "attribute_id": attribute.id,
+                            "value_ids": [
+                                Command.link(value.id) for value in attribute.value_ids
+                            ],
+                        }
+                    )
+                    for attribute in self.attribute_ids
+                ],
+            }
+        )
+        other_templates = (
+            self.line_ids.product_id.product_tmpl_id - self.product_tmpl_id
+        )
+        for line in self.line_ids:
+            product_variant = line.product_id
+            self._update_pricelist_item(product_variant)
+            self._update_supplier_info(product_variant)
+            self._move_variant_to_template(product_variant, line.attribute_value_ids)
+        merge_records(
+            self.env,
+            self.product_tmpl_id._name,
+            other_templates.ids,
+            self.product_tmpl_id.id,
+            field_spec=self._get_merge_field_spec(),
+            delete=False,
```

Comment:
Out of curiosity. Why to keep the old template?

Reply:
As a merge strategy, we chose to keep the target values. This behavior can be modified through inheritance of _get_merge_field_spec. However, after the merge, the user may need to check one of the merged products to see, for example, the previous price or any other useful data.

That’s why we archive the merged products and leave the cleanup to the user if they wish to do so.